File: ios/Plugin/GetFeedbackCapacitorPlugin.swift
import Foundation
import Capacitor
import Usabilla
import UIKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(GetFeedbackCapacitorPlugin)
public class GetFeedbackCapacitorPlugin: CAPPlugin {
    var passiveCallID: String?
    var campaignCallID: String?
    var standardEventsCallID: String?

    let getfeedback = Usabilla.self
    override public func load() {
        getfeedback.delegate = self
    }

    @objc weak var formNavigationController: UINavigationController?

    @objc func initialize(_ call: CAPPluginCall) {
        let appID = call.getString("appID") ?? ""
        getfeedback.initialize(appID: appID)
    }

    @objc func standardEvents(_ call: CAPPluginCall) {
        self.bridge?.saveCall(call)
        call.keepAlive = true
        standardEventsCallID = call.callbackId
    }

    @objc func loadFeedbackForm(_ call: CAPPluginCall) {
        let formID = call.getString("formID") ?? ""
        self.bridge?.saveCall(call)
        passiveCallID = call.callbackId
        getfeedback.loadFeedbackForm(formID)
    }

    @objc func loadFeedbackFormWithCurrentViewScreenshot(_ call: CAPPluginCall) {
        let formID = call.getString("formID") ?? ""
        self.bridge?.saveCall(call)
        passiveCallID = call.callbackId
        DispatchQueue.main.sync {
            if let rootVC = UIApplication.shared.windows.first?.rootViewController {
                let screenshot = self.takeScreenshot(view: rootVC.view)
                getfeedback.loadFeedbackForm(formID, screenshot: screenshot)
            }
        }
    }

    @objc func preloadFeedbackForms(_ call: CAPPluginCall) {
        guard let formIDs = call.options["formIDs"] as? [String] else {
            call.reject("Must provide formIDs")
            return
        }
        getfeedback.preloadFeedbackForms(withFormIDs: formIDs)
    }

    @objc func removeCachedForms(_ call: CAPPluginCall) {
        getfeedback.removeCachedForms()
    }

    @objc func sendEvent(_ call: CAPPluginCall) {
        let eventName = call.getString("eventName") ?? ""
        self.bridge?.saveCall(call)
        campaignCallID = call.callbackId
        getfeedback.sendEvent(event: eventName)
    }

    @objc func setCustomVariables(_ call: CAPPluginCall) {
        guard let variables = call.options["customVariables"] as? [String: String] else {
            call.reject("Must provide variables")
            print("ERROR : Expected customVariables as Dictionary of String [String : String]")
            return
        }
        getfeedback.customVariables = variables
    }

    @objc func setDebugEnabled(_ call: CAPPluginCall) {
        guard let debugEnabled = call.options["debugEnabled"] as? Bool else {
            call.reject("Must provide debugEnabled")
            return
        }
        getfeedback.debugEnabled = debugEnabled
    }

    @objc func loadLocalizedStringFile(_ call: CAPPluginCall) {
        let localizedStringFile = call.getString("localizedStringFile") ?? ""
        getfeedback.localizedStringFile = localizedStringFile
    }

    func takeScreenshot(view: UIView) -> UIImage {
        return getfeedback.takeScreenshot(view)!
    }

    @objc func resetCampaignData(_ call: CAPPluginCall) {
        getfeedback.resetCampaignData {
        }
    }

    @objc func dismiss(_ call: CAPPluginCall) {
        _ = getfeedback.dismiss()
    }

    @objc func getDefaultDataMasks(_ call: CAPPluginCall) {
        let str = getfeedback.defaultDataMasks
        call.resolve(["DEFAULT_DATA_MASKS": str])
    }

    @objc func setDataMasking(_ call: CAPPluginCall) {
        let masks = call.options["masks"] as? [String] ?? getfeedback.defaultDataMasks
        let maskChar = call.getString("character")?.first ?? "X"
        Usabilla.setDataMasking(masks: masks, maskCharacter: maskChar)
    }
}

extension GetFeedbackCapacitorPlugin: UsabillaDelegate {
    public func formDidLoad(form: UINavigationController) {
        formNavigationController = form
        if let rootVC = UIApplication.shared.keyWindow?.rootViewController {
            rootVC.present(formNavigationController!, animated: true, completion: nil)
        }
    }

    public func formDidFailLoading(error: UBError) {
        formNavigationController = nil
        if let callID = passiveCallID, let call = bridge?.savedCall(withID: callID) {
            call.reject("The form could not be loaded")
            bridge?.releaseCall(call)
        }
    }

    public func formDidClose(formID: String, withFeedbackResults results: [FeedbackResult], isRedirectToAppStoreEnabled: Bool) {
        var rnResults: [[String: Any]] = []
        for result in results {
            let dictionary: Dictionary = ["rating": result.rating ?? 0, "abandonedPageIndex": result.abandonedPageIndex ?? 0, "sent": result.sent] as [String: Any]
            rnResults.append(dictionary)
        }
        formNavigationController = nil

        if let callID = passiveCallID, let call = bridge?.savedCall(withID: callID) {
            call.resolve(["formId": formID, "results": rnResults, "isRedirectToAppStoreEnabled": isRedirectToAppStoreEnabled])
            bridge?.releaseCall(call)
        }
    }

    public func campaignDidClose(withFeedbackResult result: FeedbackResult, isRedirectToAppStoreEnabled: Bool) {
        let rnResult: [String: Any] = ["rating": result.rating ?? 0, "abandonedPageIndex": result.abandonedPageIndex ?? 0, "sent": result.sent] as [String: Any]
        formNavigationController = nil

        if let callID = campaignCallID, let call = bridge?.savedCall(withID: callID) {
            call.resolve(["result": rnResult, "isRedirectToAppStoreEnabled": isRedirectToAppStoreEnabled])
            bridge?.releaseCall(call)
            return
        }

        if let callEventID = standardEventsCallID, let callStandardEvent = bridge?.savedCall(withID: callEventID) {
            callStandardEvent.resolve(["result": rnResult, "isRedirectToAppStoreEnabled": isRedirectToAppStoreEnabled])
        }
    }
}

File: android/src/main/java/com/getfeedback/getfeedbackcapacitor/GetFeedbackCapacitorPlugin.java
package com.getfeedback.getfeedbackcapacitor;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.usabilla.sdk.ubform.UbConstants;
import com.usabilla.sdk.ubform.Usabilla;
import com.usabilla.sdk.ubform.UsabillaFormCallback;
import com.usabilla.sdk.ubform.sdk.entity.FeedbackResult;
import com.usabilla.sdk.ubform.sdk.form.FormClient;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "GetFeedbackCapacitor")
public class GetFeedbackCapacitorPlugin extends Plugin implements UsabillaFormCallback {

    private Usabilla getfeedback = Usabilla.INSTANCE;
    private static final String LOG_TAG = "Usabilla Ionic Bridge";
    private Fragment form;
    public static final String FRAGMENT_TAG = "passive form";
    private static final String DEFAULT_DATA_MASKS = "DEFAULT_DATA_MASKS";
    private static final String KEY_ABANDONED_PAGE_INDEX = "abandonedPageIndex";
    private static final String KEY_RATING = "rating";
    private static final String KEY_SENT = "sent";
    private static final JSArray defaultDataMasks = new JSArray(UbConstants.getDefaultDataMasks());
    private static final String defaultMaskCharacter = "X";

    private String passiveCallbackId;
    private String campaignCallbackId;
    private String standardEventsCallID;

    private BroadcastReceiver closingCampaignReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent != null) {
                final JSObject result = prepareResult(intent, FeedbackResult.INTENT_FEEDBACK_RESULT_CAMPAIGN);
                PluginCall call = bridge.getSavedCall(campaignCallbackId);
                if (call != null) {
                    call.resolve(result);
                    bridge.releaseCall(call);
                } else {
                    PluginCall callStandardEvent = bridge.getSavedCall(standardEventsCallID);
                    callStandardEvent.resolve(result);
                }
            }
        }
    };
    private BroadcastReceiver closingFormReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent != null) {
                final JSObject result = prepareResult(intent, FeedbackResult.INTENT_FEEDBACK_RESULT);
                PluginCall call = bridge.getSavedCall(passiveCallbackId);
                call.resolve(result);
                bridge.releaseCall(call);
            }
            final Activity activity = getActivity();
            if (activity instanceof FragmentActivity) {
                final FragmentManager supportFragmentManager = ((FragmentActivity) activity).getSupportFragmentManager();
                final Fragment fragment = supportFragmentManager.findFragmentByTag(FRAGMENT_TAG);
                if (fragment != null) {
                    supportFragmentManager.beginTransaction().remove(fragment).commit();
                }
            }
        }
    };

    @Override
    public void load() {
        super.load();
        LocalBroadcastManager
            .getInstance(getActivity())
            .registerReceiver(closingFormReceiver, new IntentFilter(UbConstants.INTENT_CLOSE_FORM));
        LocalBroadcastManager
            .getInstance(getActivity())
            .registerReceiver(closingCampaignReceiver, new IntentFilter(UbConstants.INTENT_CLOSE_CAMPAIGN));
    }

    @Override
    public void handleOnDestroy() {
        super.handleOnDestroy();
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(closingFormReceiver);
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(closingCampaignReceiver);
    }

    @PluginMethod
    public void initialize(@NotNull PluginCall call) {
        String appID = call.getString("appID");
        final Activity activity = getActivity();
        if (activity != null) {
            getfeedback.initialize(activity.getBaseContext(), appID);
            getfeedback.updateFragmentManager(((FragmentActivity) activity).getSupportFragmentManager());
            return;
        }
        Log.e(LOG_TAG, "Initialisation not possible. Android activity is null");
    }

    @PluginMethod(returnType = PluginMethod.RETURN_CALLBACK)
    public void standardEvents(@NotNull PluginCall call) {
        bridge.saveCall(call);
        call.setKeepAlive(true);
        standardEventsCallID = call.getCallbackId();
    }

    @PluginMethod
    public void loadFeedbackForm(@NotNull PluginCall call) {
        String formID = call.getString("formID");
        bridge.saveCall(call);
        passiveCallbackId = call.getCallbackId();
        getfeedback.loadFeedbackForm(formID, null, null, this);
    }

    @PluginMethod
    public void loadFeedbackFormWithCurrentViewScreenshot(@NotNull PluginCall call) {
        String formID = call.getString("formID");
        bridge.saveCall(call);
        passiveCallbackId = call.getCallbackId();
        final Activity activity = getActivity();
        if (activity != null) {
            final Bitmap screenshot = getfeedback.takeScreenshot(activity);
            getfeedback.loadFeedbackForm(formID, screenshot, this);
            return;
        }
        Log.e(LOG_TAG, "Loading feedback form not possible. Android activity is null");
    }

    @PluginMethod
    public void preloadFeedbackForms(@NotNull PluginCall call) throws JSONException {
        JSArray formIDs = call.getArray("formIDs");
        List<String> listformIDs = new ArrayList<>();
        for (int i = 0; i < formIDs.length(); i++) {
            listformIDs.add(formIDs.getString(i));
        }
        getfeedback.preloadFeedbackForms(listformIDs);
    }

    @PluginMethod
    public void removeCachedForms(@NotNull PluginCall call) {
        getfeedback.removeCachedForms();
    }

    @PluginMethod
    public void setCustomVariables(@NotNull PluginCall call) {
        JSObject customVars = call.getObject("customVariables");
        final HashMap<String, Object> customVariables = new HashMap<>();
        if (customVars != null) {
            for (Iterator<String> it = customVars.keys(); it.hasNext();) {
                final String key = it.next();
                Object value;
                try {
                    value = customVars.get(key);
                } catch (JSONException e) {
                    value = null;
                    e.printStackTrace();
                }
                customVariables.put(key, value);
            }
        }
        getfeedback.setCustomVariables(customVariables);
    }

    @PluginMethod
    public void sendEvent(@NotNull PluginCall call) {
        String eventName = call.getString("eventName");
        bridge.saveCall(call);
        campaignCallbackId = call.getCallbackId();
        final Activity activity = getActivity();
        if (activity != null) {
            getfeedback.sendEvent(activity.getBaseContext(), eventName);
            return;
        }
        Log.e(LOG_TAG, "Sending event to Usabilla is not possible. Android activity is null");
    }

    @PluginMethod
    public void resetCampaignData(@NotNull PluginCall call) {
        final Activity activity = getActivity();
        if (activity != null) {
            getfeedback.resetCampaignData(activity.getBaseContext());
            return;
        }
        Log.e(LOG_TAG, "Resetting Usabilla campaigns is not possible. Android activity is null");
    }

    @PluginMethod
    public boolean dismiss(@NotNull PluginCall call) {
        final Activity activity = getActivity();
        if (activity != null) {
            return getfeedback.dismiss(activity.getBaseContext());
        }
        Log.e(LOG_TAG, "Dismissing the Usabilla form is not possible. Android activity is null");
        return false;
    }

    @PluginMethod
    public void setDebugEnabled(@NonNull PluginCall call) {
        final boolean debugEnabled = call.getBoolean("debugEnabled");
        getfeedback.setDebugEnabled(debugEnabled);
    }

    @PluginMethod
    public void setDataMasking(@NonNull PluginCall call) {
        JSArray masks = call.getArray("masks", defaultDataMasks);
        String character = call.getString("character", defaultMaskCharacter);
        List<String> listMasks = null;
        try {
            listMasks = masks.toList();
        } catch (JSONException e) {
            e.printStackTrace();
        }
        getfeedback.setDataMasking(listMasks, character.charAt(0));
    }

    @PluginMethod
    public void getDefaultDataMasks(@NonNull PluginCall call) {
        final JSObject result = new JSObject();
        String[] array = UbConstants.getDefaultDataMasks().toArray(new String[0]);
        try {
            result.put(DEFAULT_DATA_MASKS, new JSArray(array));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        call.resolve(result);
    }

    @Override
    public void formLoadFail() {
        PluginCall call = bridge.getSavedCall(passiveCallbackId);
        call.reject("The form could not be loaded");
        bridge.releaseCall(call);
    }

    @Override
    public void formLoadSuccess(@NotNull FormClient formClient) {
        form = formClient.getFragment();
        final Activity activity = getActivity();
        if (activity instanceof FragmentActivity && form != null) {
            ((FragmentActivity) activity).getSupportFragmentManager()
                .beginTransaction()
                .replace(android.R.id.content, form, FRAGMENT_TAG)
                .commit();
            form = null;
        }
    }

    private JSObject prepareResult(Intent intent, String feedbackResultType) {
        final JSObject result = new JSObject();
        final JSObject resultData = new JSObject();
        String res = (feedbackResultType == FeedbackResult.INTENT_FEEDBACK_RESULT) ? "results" : "result";
        resultData.put(res, getResult(intent, feedbackResultType));
        result.put("completed", resultData);
        return result;
    }

    private JSObject getResult(Intent intent, String feedbackResultType) {
        final FeedbackResult res = intent.getParcelableExtra(feedbackResultType);
        final JSObject result = new JSObject();
        result.put(KEY_RATING, res.getRating());
        result.put(KEY_ABANDONED_PAGE_INDEX, res.getAbandonedPageIndex());
        result.put(KEY_SENT, res.isSent());
        return result;
    }

    @Override
    public void mainButtonTextUpdated(@NotNull String s) {
        // To fill with handling required when the main button text changes
    }
}
